_posts/2021-03-23-keccak-optimizer-bug.md

```diff
@@ -0,0 +1,89 @@
+---
+layout: post
+published: true
+title:  "Solidity Keccak Caching bug"
+date:   2021-03-23
+author: Solidity Team
+category: Security Alerts
+---
+
+On March 20, 2021, a bug in Solidity's bytecode optimizer was found. The bug is fixed with [version
+0.8.3](https://github.com/ethereum/solidity/releases/tag/v0.8.3) released on March 23, 2021. **The
+bug is present in all prior versions of Solidity.**
+
+We assigned the bug a severity level of "medium".
+
+## Technical Details of the Bug
+
+**Summary**: The bytecode optimizer incorrectly re-used previously evaluated Keccak-256 hashes. You
+are unlikely to be affected if you do not compute Keccak-256 hashes in inline assembly.
+
+Solidity's bytecode optimizer has a step that can compute Keccak-256 hashes, if the contents of the
+memory are known during compilation time. This step also has a mechanism to determine that two
+Keccak-256 hashes are equal even if the values in memory are not known during compile time. This
+mechanism had a bug where, Keccak-256 of the same memory content, but different sizes were
+considered equal. The following is a example that illustrates the bug:
+
+```solidity
+contract C {
+  uint[] data;
+
+  function bug() public{
+    assembly {
+      // Sets data.length = 2
+      sstore(0, 2)
+      mstore(0, 0)
+      // data[0] = 234
+      sstore(keccak256(0, 32), 234)
+      // A bug in the caching mechanism previously caused keccak256(0, 23) to be the same as
+      // keccak256(0, 32), leading to data[1] == 123` being true.
+      sstore(add(keccak256(0, 23), 1), 123)
```

Comment:
I think this is a bit distracting - shouldn't we focus on a simple example that just shows exactly the bug - especially as the first example?

Reply:
I think for better readability it is good to draw attention to an example where memory contents are not known at compile-time but are considered equal due to the bug. So perhaps the `string` example where the keccak input is a function parameter?

Reply:
I changed this example to a much simpler one.